### ISSUE TITLE
map_entry is incorrectly suggested when map is used between contains_key and insert

### ISSUE THREAD

**Issue by khyperia** [rust-lang/rust-clippy#6196]
Related issue: https://github.com/rust-lang/rust-clippy/issues/5696

Reproducing code:

```rust
use std::collections::HashMap;

// don't want to call this if contains_key==true
fn really_expensive_function(_map: &HashMap<u32, u32>) {}

fn main() {
    let mut map = HashMap::new();
    let key = 5;
    if !map.contains_key(&key) {
        really_expensive_function(&map);
        map.insert(key, 10);
    }
    // This doesn't compile:
    /*
    use std::collections::hash_map::Entry;
    match map.entry(key) {
        Entry::Vacant(entry) => {
            really_expensive_function(&map);
            entry.insert(10);
        }
        _ => (),
    }
    */
}
```

produces

```
warning: usage of `contains_key` followed by `insert` on a `HashMap`
  --> src/main.rs:9:5
   |
9  | /     if !map.contains_key(&key) {
10 | |         really_expensive_function(&map);
11 | |         map.insert(key, 10);
12 | |     }
   | |_____^ consider using `map.entry(key)`
```

However, the commented-out code (which is what I believe clippy is suggesting) fails to compile with

```
error[E0502]: cannot borrow `map` as immutable because it is also borrowed as mutable
  --> src/main.rs:17:39
   |
15 |     match map.entry(key) {
   |           --- mutable borrow occurs here
16 |         Entry::Vacant(entry) => {
17 |             really_expensive_function(&map);
   |                                       ^^^^ immutable borrow occurs here
18 |             entry.insert(10);
   |             ----- mutable borrow later used here
```

I believe the correct behavior here is to not emit a suggestion at all.

### Meta

- `cargo clippy -V`: clippy 0.0.212 (18bf6b4 2020-10-07)
- `rustc -Vv`:
  ```
  rustc 1.47.0 (18bf6b4f0 2020-10-07)
  binary: rustc
  commit-hash: 18bf6b4f01a6feaf7259ba7cdae58031af1b7b39
  commit-date: 2020-10-07
  host: x86_64-unknown-linux-gnu
  release: 1.47.0
  LLVM version: 11.0
  ```

**Comment by giraffate** [comment]
related issue: https://github.com/rust-lang/rust-clippy/issues/4674

**Comment by khyperia** [comment]
> related issue: #4674

Looks like this (my issue) is a duplicate, feel free to close this if you'd like.
